src/sm.c

```diff
@@ -0,0 +1,480 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+#include <stdlib.h>
+#include <stdint.h>
+#include <inttypes.h>
+
+#include "windows.h"
+
+#include "azure_macro_utils/macro_utils.h"
+
+#include "azure_c_util/xlogging.h"
+#include "azure_c_util/gballoc.h"
+#include "azure_c_util/interlocked_hl.h"
+
+#include "azure_c_util/sm.h"
+
+#define SM_STATE_VALUES             \
+    SM_CREATED,                     \
+    SM_OPENING,                     \
+    SM_OPENED,                      \
+    SM_OPENED_DRAINING_TO_BARRIER,  \
+    SM_OPENED_DRAINING_TO_CLOSE,    \
+    SM_OPENED_BARRIER,              \
+    SM_CLOSING                      \
+
+MU_DEFINE_ENUM(SM_STATE, SM_STATE_VALUES)
+
+#define SM_STATE_MASK       127
+#define SM_STATE_INCREMENT  256 /*at every state change, the state is incremented by this much*/
+
+#define SM_CLOSE_BIT 128
```

Comment:
>128 [](start = 21, length = 3)

Consider using 1 << 7 (like we used to do in automotive to be explicit about the bit part (and avoid such a monkey like me from changing this to a bad value if let's say I'd have to just change the bit :-))

Reply:
Same applied to the mask SM_STATE_MASK

---
In reply to: [423409725](https://github.com/Azure/azure-c-util/pull/37#discussion_r423409725) [](ancestors = 423409725)

Reply:
Alright, used shifts... 

---
In reply to: [423409833](https://github.com/Azure/azure-c-util/pull/37#discussion_r423409833) [](ancestors = 423409833,423409725)